tls/s2n_handshake_io.c

```diff
@@ -823,19 +825,22 @@ static int handshake_read_io(struct s2n_connection *conn)
      */
     S2N_ERROR_IF(record_type == TLS_APPLICATION_DATA, S2N_ERR_BAD_MESSAGE);
     if (record_type == TLS_CHANGE_CIPHER_SPEC) {
+        /* TLS1.2 should not receive unexpected change cipher spec messages, but TLS1.3 might. */
+        S2N_ERROR_IF(!IS_TLS13_HANDSHAKE(conn) && EXPECTED_RECORD_TYPE(conn) != TLS_CHANGE_CIPHER_SPEC, S2N_ERR_BAD_MESSAGE);
```

Comment:
Using unary negation and binary operators with macros is asking for trouble :) The macro could change later and just by adding parentheses or become a multi-statement clause, alter the intended effect here. I'd either add some gratuitous parens here for safety, or just use a regular series of if blocks :)

Reply:
I split out the unary negation into an separate if.

We probably didn't have a test for early CCS, because it was possible in the existing code. I added some new unit tests (s2n_tls12_handshake_test), and they fail when I remove this error check.